helix-term/src/commands.rs

```diff
@@ -701,27 +701,90 @@ pub fn split_selection_on_newline(cx: &mut Context) {
     doc.set_selection(view.id, selection);
 }
 
-// search: searches for the first occurence in file, provides a prompt
-// search_next: reuses the last search regex and searches for the next match. The next match becomes the main selection.
 // -> we always search from after the cursor.head
 // TODO: be able to use selection as search query (*/alt *)
-// I'd probably collect all the matches right now and store the current index. The cache needs
-// wiping if input happens.
 
-fn _search(doc: &mut Document, view: &mut View, contents: &str, regex: &Regex, extend: bool) {
-    let text = doc.text();
+use helix_core::search::Searcher;
+
+pub fn search(cx: &mut Context) {
+    let (view, doc) = cx.current();
+
+    // TODO: could probably share with select_on_matches?
+
+    let view_id = view.id;
+    let snapshot = doc.selection(view_id).clone();
+
+    let prompt = Prompt::new(
+        "search:".to_string(),
+        |input: &str| Vec::new(), // this is fine because Vec::new() doesn't allocate
+        move |editor: &mut Editor, input: &str, event: PromptEvent| {
+            match event {
+                PromptEvent::Abort => {
+                    let (view, doc) = editor.current();
+                    doc.set_selection(view.id, snapshot.clone());
+                }
+                PromptEvent::Validate => {
+                    // TODO: push_jump to store selection just before jump
+                }
+                PromptEvent::Update => {
+                    // skip empty input, TODO: trigger default
+                    if input.is_empty() {
+                        return;
+                    }
+
+                    match Searcher::new(input) {
```

Comment:
This seemed to be blocking the input, even when I search (debug) "hello world" every character takes like few seconds and I need to wait. Maybe we should spawn it as a blocking task?